### PR TITLE
Add support for INSERT INTO SELECT statement

### LINKED ISSUE #1158: INSERT INTO SELECT
`Insert into select` currently doesn't insert anything and does not generate code to do so for a few reasons.
1) InsertTranslator doesn't call prepare on its children if such children exist.

2) The InsertPlanNode needs to have DerivedValueExpressions in its values_ vector to signify tuples that are derived from its children for insertion. This should be done in `plan_generator.cpp`. Otherwise, there is nothing in the values_ vector during `INSERT INTO SELECT` and the insert translator generates nothing.

3) Also, in order for the children of the insert node to pass up tuples the InsertSelect node needs to request the required columns from the descendant scan node [here](https://github.com/cmu-db/terrier/blob/d60db2a543eed2c0463d2fae02eb4bc39628b4f8/src/optimizer/input_column_deriver.cpp#L233). Currently this just passes down required columns from above which is incorrect. You can't request columns from an Insert node. Also, at this point we must generate the required columns to request from the children. In the event that we have a `INSERT INTO yyy SELECT * FROM xxx;`, we can easily just query all the columns oids in xxx and request those columns. The tableoid of xxx is stored in the child of the concerned InsertSelect node.

4) The information stored in InsertSelect is incomplete. While the vanilla Insert physical operator node has [this](https://github.com/cmu-db/terrier/blob/e525b6ddde0598019d2a8d58e2d8907d0d3051b1/src/include/optimizer/physical_operators.h#L896) that allows us to know which columns we are inserting into, the InsertSelect operator has no equivalent even though this information is necessary.

### PR DESCRIPTION
**Pull request by jkosh44**
## Description
The implementation is modeled very closely after normal inserts, with two major difference.
1. In the optimizer the InsertSelect node has an output of the child Select node. This output is then pushed down to the child select node in the `InputColumnDeriver`.
2. In the translator there are two differences.
  a. At the start if it's an INSERT INTO SELECT it will prepare the child select translator. If it's a normal INSERT it will prepare the insert values.
 b. When setting the projected row if it's an INSERT INTO SELECT it will set it from the child output, if it's a normal INSERT it will set it from the values.

Fixes #1158

## Remaining tasks
None

## Performance
No performance testing was done for this PR

## Further work
N/A

### CODE REVIEW

**Comment by jkosh44** [comment]
If this is merged it should probably wait until after #1394

File: src/include/traffic_cop/traffic_cop_util.h [hunk 1]
```diff
@@ -7,6 +7,8 @@
 #include "common/managed_pointer.h"
 #include "network/network_defs.h"
 #include "optimizer/optimize_result.h"
+#include "optimizer/property_set.h"
```
Comment (lmwnshn) [inline src/include/traffic_cop/traffic_cop_util.h:10]: forward declare this

File: src/include/traffic_cop/traffic_cop_util.h [hunk 2]
```diff
@@ -7,6 +7,8 @@
...
+#include "parser/select_statement.h"
```
Comment (lmwnshn) [inline src/include/traffic_cop/traffic_cop_util.h:11]: forward declare this

**Review by lmwnshn — CHANGES_REQUESTED** [review]
LGTM pending minor nits

File: test/execution/compiler_test.cpp [hunk 3]
```diff
@@ -3082,16 +3083,16 @@ TEST_F(CompilerTest, DISABLED_InsertIntoSelectWithParamTest) {
     executable->Run(common::ManagedPointer(exec_ctx), MODE);
 
     // Pipeline Units
-    auto pipeline = executable->GetPipelineOperatingUnits();
-    EXPECT_FALSE(true);
+    /*auto pipeline = executable->GetPipelineOperatingUnits();
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply test/execution/compiler_test.cpp:3086]: Honestly I have no idea what this is, but the test fails with this not commented out.

File: test/optimizer/operator_transformer_test.cpp [hunk 4]
```diff
@@ -253,7 +253,7 @@ TEST_F(OperatorTransformerTest, InsertStatementSimpleTest) {
 // NOLINTNEXTLINE
 TEST_F(OperatorTransformerTest, InsertStatementSelectTest) {
   OPTIMIZER_LOG_DEBUG("Parsing sql query");
-  std::string insert_sql = "INSERT INTO A (A1) SELECT B1 FROM B WHERE B1 > 0";
+  std::string insert_sql = "INSERT INTO A (A1, A2) SELECT B1, B2 FROM B WHERE B1 > 0";
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply test/optimizer/operator_transformer_test.cpp:256]: Sorry, after my first pass INSERT INTO SELECT only worked if you specified every column which is why I changed it. I meant to change it back when I got it working fully but forgot. I'm going to change it back.

File: src/binder/bind_node_visitor.cpp [hunk 5]
```diff
@@ -310,147 +310,39 @@ void BindNodeVisitor::Visit(common::ManagedPointer<parser::InsertStatement> node
   context_->AddRegularTable(catalog_accessor_, db_oid_, table->GetNamespaceName(), table->GetTableName(),
                             table->GetTableName());
 
+  auto binder_table_data = context_->GetTableMapping(table->GetTableName());
+  const auto &table_schema = std::get<2>(*binder_table_data);
+
+  // Perform input validation and parsing of strings into dates.
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply src/binder/bind_node_visitor.cpp]: Done

File: src/execution/compiler/operator/insert_translator.cpp [hunk 6]
```diff
@@ -60,17 +79,39 @@ void InsertTranslator::PerformPipelineWork(WorkContext *context, FunctionBuilder
...
-    GenSetTablePR(function, context, idx);
-    // var insert_slot = @tableInsert(&inserter)
-    GenTableInsert(function);
-    function->Append(GetCodeGen()->ExecCtxAddRowsAffected(GetExecutionContext(), 1));
-    const auto &index_oids = GetPlanAs<planner::InsertPlanNode>().GetIndexOids();
-    for (const auto &index_oid : index_oids) {
-      GenIndexInsert(context, function, index_oid);
+  switch (plan.GetInsertType()) {
+    case parser::InsertType::SELECT: {
+      // var insert_pr = @getTablePR(&inserter)
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply src/execution/compiler/operator/insert_translator.cpp]: I managed to abstract it out, let me know what you think of my approach.

File: src/include/binder/bind_node_visitor.h [hunk 7]
```diff
@@ -4,6 +4,7 @@
 #include <string>
 #include <vector>
 
+#include "binder/binder_context.h"
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply src/include/binder/bind_node_visitor.h]: Done

File: src/include/execution/compiler/operator/insert_translator.h [hunk 8]
```diff
@@ -86,8 +86,11 @@ class InsertTranslator : public OperatorTranslator, public PipelineDriver {
   // Gets the projected row pointer that we will fill in with values to insert.
   void GetInsertPR(FunctionBuilder *builder) const;
 
-  // Sets the values in the projected row which we will use to insert into the table.
-  void GenSetTablePR(FunctionBuilder *builder, WorkContext *context, uint32_t idx) const;
+  // Sets the values in the projected row for a value insert which we will use to insert into the table.
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply src/include/execution/compiler/operator/insert_translator.h]: Done

File: src/include/optimizer/logical_operators.h [hunk 9]
```diff
@@ -804,6 +806,11 @@ class LogicalInsertSelect : public OperatorNodeContents<LogicalInsertSelect> {
    */
   const catalog::table_oid_t &GetTableOid() const { return table_oid_; }
 
+  /**
+   * @return OIDs of the columns that this operator is inserting into for the target table
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply src/include/optimizer/logical_operators.h]: Done

File: src/optimizer/plan_generator.cpp [hunk 10]
```diff
@@ -851,14 +852,20 @@ void PlanGenerator::Visit(const InsertSelect *op) {
...
+  auto builder = planner::InsertPlanNode::Builder();
+  builder.SetOutputSchema(std::move(output_schema))
+      .SetPlanNodeId(GetNextPlanNodeID())
+      .SetDatabaseOid(op->GetDatabaseOid())
+      .SetTableOid(op->GetTableOid())
+      .SetIndexOids(std::move(indexes))
+      .SetInsertType(parser::InsertType::SELECT)
+      .AddChild(std::move(children_plans_[0]));
+  // This is based on what Peloton does/did with query_to_operator_transformer.cpp
+  NOISEPAGE_ASSERT(!op->GetColumns().empty(), "Transformer should added columns");
```
[the remark this answers is not in the document]
Comment (jkosh44) [reply src/optimizer/plan_generator.cpp]: Done

File: test/optimizer/operator_transformer_test.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (jkosh44) [reply test/optimizer/operator_transformer_test.cpp]: Reverted

File: src/include/traffic_cop/traffic_cop_util.h (hunk 1 above)
Comment (jkosh44) [reply src/include/traffic_cop/traffic_cop_util.h]: Done.... and now I know what forward declare means...

File: src/include/traffic_cop/traffic_cop_util.h (hunk 2 above)
Comment (jkosh44) [reply src/include/traffic_cop/traffic_cop_util.h]: Done

File: src/execution/compiler/operator/insert_translator.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (lmwnshn) [reply src/execution/compiler/operator/insert_translator.cpp]: sure, looks decent

File: test/execution/compiler_test.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (lmwnshn) [reply test/execution/compiler_test.cpp:3086]: ok

File: src/include/traffic_cop/traffic_cop_util.h (hunk 1 above)
Comment (lmwnshn) [reply src/include/traffic_cop/traffic_cop_util.h]: generally it is something worth doing in commonly included .h files that could result in a huge include blowup

**Review by lmwnshn — APPROVED** [review]
LGTM! if passes ci. thanks for quick fixes

**Comment by lmwnshn** [comment]
@jkosh44 one of the changes can be taken as a `const &`

**Pull request merged**